js/listdates-min.js

```diff
@@ -1,53 +1,113 @@
         // returns '25 days ago' or '19 hours ago'
         const shortDuration = eventDate.fromNow();
 
+        // long duration text
         // https://github.com/EvanHahn/HumanizeDuration.js
         // returns '31 years, 8 months'
         let longDuration = humanizeDuration(diff, { largest: 2, round: true });
 
-        // add human-readable duration text
+        // add humanizing modifiers to long duration text
         if (diff < 0) {
-            // dates in the past
+            // this date is in the past
             longDuration = longDuration + " ago";
         } else {
-            // dates in the future
+            // this date is in the future
             longDuration = "in " + longDuration;
         }
 
-        // decide which text to use based on how long the duration is
+        // decide which duration text to use based on how long the duration is,
+        // because the short duration text doesn't work well beyond a month
         if (Math.abs(diff) < millisecondsMonth) {
             // durations shorter than a month
-            durationText = shortDuration;
+            dateArray[i].duration = shortDuration;
         } else {
             // durations longer than a month
-            durationText = longDuration;
+            dateArray[i].duration = longDuration;
         }
 
-        // construct the HTML table row
+        // format date to be human-readable (ll = Sep 4, 1986)
+        dateArray[i].calendar = eventDate.calendar(null, {sameElse: 'll'});
+
+        // add month key for sorting (L = 09/04/1986)
+        dateArray[i].month = eventDate.calendar(null, {sameElse: 'L'});
+
+        // add diff key for sorting
+        dateArray[i].diff = diff;
+    }
+
+    return dateArray;
+};
+
+// Convert date array into HTML
+const buildHTML = function(dateArray) {
+    'use strict';
+    let content = "";
+
+    // add a table row for each item in the date array
+    for (let i = 0; i < dateArray.length; i++) {
         content += "<tr>" +
-                   "<td>" + dates[i].event + "</td>" +
-                   "<td class='num'>" + eventDate.calendar(null, {sameElse: 'll'}) + "</td>" +
-                   "<td class='num'>" + durationText + "</td>" +
+                   "<td>" + dateArray[i].event + "</td>" +
+                   "<td class='num'>" + dateArray[i].calendar + "</td>" +
+                   "<td class='num'>" + dateArray[i].duration + "</td>" +
                    "</tr>";
 
     }
+
     return content;
-};
+}
+
+// Sort the date array and rebuild the HTML using the new sort
+const sortBy = function() {
+    'use strict';
+    const sortedDates = sortByKey(dates, this.dataset.sortBy);
+    datelistContent.innerHTML = buildHTML(sortedDates);
+}
+
+// Add event listeners to sort buttons
+// There's gotta be a better way, right?
+btnSortEvent.addEventListener('click', sortBy);
+btnSortMonth.addEventListener('click', sortBy);
+btnSortDuration.addEventListener('click', sortBy);
```

Comment:
So, you didn't mention this part, which I expected to get flagged. Is this dumb? Is there a better way?

Reply:
I guess I passed by it because you didn't exceed like, 5 repetitions. In jQuery I would do something like `$('.btn-sort-by').on('click', sortBy)`. I'm not sure how you do that with VanillaJS though. The other thing you could do is attach the listener to a parent object and and use `target` vs `currentTarget` to identify the correct button.

I'm not sure where `this.dataset` is though. What is that?

Reply:
To quote Sandi Metz - "duplication is far cheaper than the wrong abstraction". So that's why I didn't think much of it. https://www.sandimetz.com/blog/2016/1/20/the-wrong-abstraction

This seems like a good candidate for deduplication since the buttons are largely going to be the same; but at the same time, it's unlikely that the number of buttons is going to change and the cost of adding one more sorting button is extremely low compared to the mental cost of figuring out a clever class-based solution or looping solution.

Reply:
Hahaha, to put it another way, the cognitive load was SO low that I didn't even notice the code. :P

Reply:
I'd say go with whatever reads better to whoever will be maintaining it :)

Reply:
this.dataset is, as far as I can tell, a clever new JS technique to access the `data-*` attributes on an element. So `this.dataset.sortBy` says "Grab the `data-sort-by` attribute"

And Chuck, I totally get what you're saying about premature optimization. My gut says I should be selecting the buttons as a group, however, and after all this discussion, I realized why: I'm going to be doing more things to them later. Right now, all I do is add a click event. But later I'll be adding/removing classes and stuff, so it makes sense to have a group selector.